File: src/lib/data/mock-data.ts
/**
 * 文件名: mock-data.ts
 * 功能描述: 虚拟经营数据库，包含4城市×3产品线×3月份共36条销售记录，
 *           以及基于关键词匹配的查询引擎
 * 主要导出: salesData（数据数组）, querySalesData（查询函数）, formatRevenue（格式化工具）
 */

import type { SalesRecord, QueryResult, QueryMeta, ChartData } from '@/types';

/**
 * 虚拟经营数据库
 * 覆盖北京、上海、深圳、广州四个城市，三条产品线，2025年Q1（1-3月）
 */
export const salesData: SalesRecord[] = [
  // ===== 北京 =====
  { id: 1,  city: '北京', productLine: '企业云服务',   revenue: 1580000, quantity: 42,  month: '1月', year: 2025 },
  { id: 2,  city: '北京', productLine: '企业云服务',   revenue: 1720000, quantity: 45,  month: '2月', year: 2025 },
  { id: 3,  city: '北京', productLine: '企业云服务',   revenue: 1650000, quantity: 43,  month: '3月', year: 2025 },
  { id: 4,  city: '北京', productLine: '数据分析平台', revenue: 980000,  quantity: 28,  month: '1月', year: 2025 },
  { id: 5,  city: '北京', productLine: '数据分析平台', revenue: 1050000, quantity: 30,  month: '2月', year: 2025 },
  { id: 6,  city: '北京', productLine: '数据分析平台', revenue: 1120000, quantity: 32,  month: '3月', year: 2025 },
  { id: 7,  city: '北京', productLine: '智能办公套件', revenue: 620000,  quantity: 156, month: '1月', year: 2025 },
  { id: 8,  city: '北京', productLine: '智能办公套件', revenue: 680000,  quantity: 168, month: '2月', year: 2025 },
  { id: 9,  city: '北京', productLine: '智能办公套件', revenue: 710000,  quantity: 175, month: '3月', year: 2025 },
  // ===== 上海 =====
  { id: 10, city: '上海', productLine: '企业云服务',   revenue: 1920000, quantity: 52,  month: '1月', year: 2025 },
  { id: 11, city: '上海', productLine: '企业云服务',   revenue: 2050000, quantity: 55,  month: '2月', year: 2025 },
  { id: 12, city: '上海', productLine: '企业云服务',   revenue: 1980000, quantity: 53,  month: '3月', year: 2025 },
  { id: 13, city: '上海', productLine: '数据分析平台', revenue: 1250000, quantity: 35,  month: '1月', year: 2025 },
  { id: 14, city: '上海', productLine: '数据分析平台', revenue: 1380000, quantity: 38,  month: '2月', year: 2025 },
  { id: 15, city: '上海', productLine: '数据分析平台', revenue: 1420000, quantity: 40,  month: '3月', year: 2025 },
  { id: 16, city: '上海', productLine: '智能办公套件', revenue: 850000,  quantity: 210, month: '1月', year: 2025 },
  { id: 17, city: '上海', productLine: '智能办公套件', revenue: 920000,  quantity: 225, month: '2月', year: 2025 },
  { id: 18, city: '上海', productLine: '智能办公套件', revenue: 880000,  quantity: 218, month: '3月', year: 2025 },
  // ===== 深圳 =====
  { id: 19, city: '深圳', productLine: '企业云服务',   revenue: 1350000, quantity: 38,  month: '1月', year: 2025 },
  { id: 20, city: '深圳', productLine: '企业云服务',   revenue: 1480000, quantity: 40,  month: '2月', year: 2025 },
  { id: 21, city: '深圳', productLine: '企业云服务',   revenue: 1520000, quantity: 41,  month: '3月', year: 2025 },
  { id: 22, city: '深圳', productLine: '数据分析平台', revenue: 820000,  quantity: 24,  month: '1月', year: 2025 },
  { id: 23, city: '深圳', productLine: '数据分析平台', revenue: 890000,  quantity: 26,  month: '2月', year: 2025 },
  { id: 24, city: '深圳', productLine: '数据分析平台', revenue: 950000,  quantity: 28,  month: '3月', year: 2025 },
  { id: 25, city: '深圳', productLine: '智能办公套件', revenue: 560000,  quantity: 140, month: '1月', year: 2025 },
  { id: 26, city: '深圳', productLine: '智能办公套件', revenue: 590000,  quantity: 148, month: '2月', year: 2025 },
  { id: 27, city: '深圳', productLine: '智能办公套件', revenue: 630000,  quantity: 158, month: '3月', year: 2025 },
  // ===== 广州 =====
  { id: 28, city: '广州', productLine: '企业云服务',   revenue: 1100000, quantity: 30,  month: '1月', year: 2025 },
  { id: 29, city: '广州', productLine: '企业云服务',   revenue: 1180000, quantity: 32,  month: '2月', year: 2025 },
  { id: 30, city: '广州', productLine: '企业云服务',   revenue: 1250000, quantity: 34,  month: '3月', year: 2025 },
  { id: 31, city: '广州', productLine: '数据分析平台', revenue: 680000,  quantity: 20,  month: '1月', year: 2025 },
  { id: 32, city: '广州', productLine: '数据分析平台', revenue: 720000,  quantity: 22,  month: '2月', year: 2025 },
  { id: 33, city: '广州', productLine: '数据分析平台', revenue: 780000,  quantity: 24,  month: '3月', year: 2025 },
  { id: 34, city: '广州', productLine: '智能办公套件', revenue: 450000,  quantity: 112, month: '1月', year: 2025 },
  { id: 35, city: '广州', productLine: '智能办公套件', revenue: 480000,  quantity: 120, month: '2月', year: 2025 },
  { id: 36, city: '广州', productLine: '智能办公套件', revenue: 520000,  quantity: 130, month: '3月', year: 2025 },
];

/**
 * 格式化金额显示
 * @param value - 金额数值（元）
 * @returns 格式化后的字符串，如 "158.00 万元" 或 "1,234 元"
 */
export function formatRevenue(value: number): string {
  if (value >= 10000) {
    return (value / 10000).toFixed(2) + ' 万元';
  }
  return value.toLocaleString() + ' 元';
}

/**
 * 生成模拟的查询元信息（耗时、Token数、时间戳）
 * @returns QueryMeta 对象
 */
function generateMeta(): QueryMeta {
  return {
    durationMs: Math.floor(Math.random() * 1300) + 1200, // 1200~2500ms
    tokenCount: Math.floor(Math.random() * 700) + 800,   // 800~1500 tokens
    timestamp: Date.now(),
  };
}

/**
 * 根据用户问题关键词查询销售数据
 * 支持三种查询模式：
 *   1. 全局产品线汇总（不含城市关键词）
 *   2. 指定城市的产品线明细
 *   3. 默认整体概况
 *
 * @param question - 用户输入的问题文本
 * @returns 结构化查询结果（含摘要、表格、洞察、图表数据）
 */
export function querySalesData(question: string): QueryResult {
  const q = question.toLowerCase();

  // 模式1：各产品线销售情况（不涉及具体城市）
  const cityNames = ['北京', '上海', '深圳', '广州'];
  const isGlobalQuery = q.includes('产品线') && !cityNames.some(c => q.includes(c));

  if (isGlobalQuery) {
    const productLines = ['企业云服务', '数据分析平台', '智能办公套件'];

    // 构建表格行
    const rows = productLines.map((pl) => {
      const records = salesData.filter((r) => r.productLine === pl);
      const totalRevenue = records.reduce((sum, r) => sum + r.revenue, 0);
      const totalQuantity = records.reduce((sum, r) => sum + r.quantity, 0);
      return [pl, formatRevenue(totalRevenue), totalQuantity.toString(), records.length.toString()];
    });

    // 构建图表数据
    const chartData: ChartData = {
      title: '各产品线营收与销量对比',
      categories: productLines,
      series: [
        {
          name: '总营收(万元)',
          values: productLines.map((pl) => {
            const total = salesData.filter(r => r.productLine === pl).reduce((s, r) => s + r.revenue, 0);
            return Number((total / 10000).toFixed(2));
          }),
          color: 'primary',
          unit: '万元',
        },
        {
          name: '总销量',
          values: productLines.map((pl) =>
            salesData.filter(r => r.productLine === pl).reduce((s, r) => s + r.quantity, 0)
          ),
          color: 'secondary',
        },
      ],
    };

    return {
      summary: '以下是 2025 年 Q1（1-3月）各产品线的销售汇总数据：',
      table: { headers: ['产品线', '总营收', '总销量', '记录数'], rows },
      insights: [
        '企业云服务是营收最高的产品线，Q1 总营收占比约 52%',
        '数据分析平台增长势头良好，3月环比增长明显',
        '智能办公套件销量最大，但客单价较低，可考虑推出高级版本',
      ],
      chartData,
      meta: generateMeta(),
    };
  }

  // 模式2：指定城市的产品线明细
  const matchedCity = cityNames.find((c) => q.includes(c));

  if (matchedCity) {
    const cityRecords = salesData.filter((r) => r.city === matchedCity);
    const productLines = [...new Set(cityRecords.map((r) => r.productLine))];

    // 构建表格行
    const rows = productLines.map((pl) => {
      const records = cityRecords.filter((r) => r.productLine === pl);
      const totalRevenue = records.reduce((sum, r) => sum + r.revenue, 0);
      const totalQuantity = records.reduce((sum, r) => sum + r.quantity, 0);
      const months = records.map((r) => r.month).join('、');
      return [pl, formatRevenue(totalRevenue), totalQuantity.toString(), months];
    });

    const totalCityRevenue = cityRecords.reduce((sum, r) => sum + r.revenue, 0);

    // 构建图表数据
    const chartData: ChartData = {
      title: `${matchedCity} 各产品线营收对比`,
      categories: productLines,
      series: [
        {
          name: '营收(万元)',
          values: productLines.map((pl) => {
            const total = cityRecords.filter(r => r.productLine === pl).reduce((s, r) => s + r.revenue, 0);
            return Number((total / 10000).toFixed(2));
          }),
          color: 'primary',
          unit: '万元',
        },
        {
          name: '销量',
          values: productLines.map((pl) =>
            cityRecords.filter(r => r.productLine === pl).reduce((s, r) => s + r.quantity, 0)
          ),
          color: 'secondary',
        },
      ],
    };

    return {
      summary: `以下是 ${matchedCity} 2025 年 Q1 各产品线的销售明细：`,
      table: { headers: ['产品线', '营收', '销量', '覆盖月份'], rows },
      insights: [
        `${matchedCity} Q1 总营收为 ${formatRevenue(totalCityRevenue)}`,
        `企业云服务在${matchedCity}表现最强，建议持续加大投入`,
        q.includes('产品')
          ? `${matchedCity}各产品线发展较为均衡，无明显短板`
          : `${matchedCity}市场潜力较大，可考虑增加资源投入`,
      ],
      chartData,
      meta: generateMeta(),
    };
  }

  // 模式3：默认回答 —— 整体经营概况
  const totalRevenue = salesData.reduce((sum, r) => sum + r.revenue, 0);
  const totalQuantity = salesData.reduce((sum, r) => sum + r.quantity, 0);
  const cityCount = new Set(salesData.map((r) => r.city)).size;

  return {
    summary: '根据当前经营数据库，以下是整体经营概况：',
    table: {
      headers: ['指标', '数值'],
      rows: [
        ['Q1 总营收', formatRevenue(totalRevenue)],
        ['Q1 总销量', totalQuantity.toString() + ' 单'],
        ['覆盖城市', cityCount + ' 个'],
        ['产品线', '3 条'],
        ['数据周期', '2025年1月 - 3月'],
      ],
    },
    insights: [
      '您可以问我具体城市或产品线的销售情况',
      '例如："北京的产品线收入情况"、"深圳的产品销售情况"',
      '目前数据覆盖北京、上海、深圳、广州四个城市',
    ],
    meta: generateMeta(),
  };
}

File: backend/data/mock_data.py
"""
虚拟数据库

模拟经营数据的存储和查询
返回格式与前端 QueryResult 类型完全匹配
"""

from typing import Optional


class MockDatabase:
    """虚拟数据库类"""

    def __init__(self):
        """初始化虚拟数据"""
        # 产品线数据
        self.product_data = {
            "产品线A": {"目标": 80, "实际": 100, "达成率": 125},
            "产品线B": {"目标": 90, "实际": 85, "达成率": 94},
            "产品线C": {"目标": 70, "实际": 78, "达成率": 111},
        }

        # 城市数据
        self.city_data = {
            "北京": {"销售额": 120, "订单量": 3500, "客单价": 343},
            "上海": {"销售额": 150, "订单量": 4200, "客单价": 357},
            "广州": {"销售额": 90, "订单量": 2800, "客单价": 321},
            "深圳": {"销售额": 110, "订单量": 3100, "客单价": 355},
        }

        # 部门数据
        self.department_data = {
            "销售部": {"员工数": 25, "业绩": 470, "人均产出": 18.8},
            "市场部": {"员工数": 15, "业绩": 280, "人均产出": 18.7},
            "运营部": {"员工数": 20, "业绩": 350, "人均产出": 17.5},
        }

    def query_data(self, message: str) -> dict:
        """
        根据消息查询数据，返回前端 QueryResult 格式

        Returns:
            { summary, insights, chartData?, meta }
        """
        if "产品线" in message or "产品" in message:
            return self._query_product_data(message)
        elif "城市" in message or "北京" in message or "上海" in message:
            return self._query_city_data(message)
        elif "部门" in message or "销售" in message or "市场" in message:
            return self._query_department_data(message)
        else:
            return self._default_response(message)

    def _query_product_data(self, message: str) -> dict:
        """查询产品线数据"""
        summary = "根据查询结果，各产品线销售情况如下：\n\n"
        for name, data in self.product_data.items():
            summary += f"- {name}：目标 {data['目标']}万，实际 {data['实际']}万，达成率 {data['达成率']}%\n"

        insights = [
            "产品线A达成率最高，达到125%",
            "产品线B未达标，建议关注原因",
            "整体产品线平均达成率为110%"
        ]

        chart_data = {
            "title": "产品线目标与实际对比（万元）",
            "categories": list(self.product_data.keys()),
            "series": [
                {
                    "name": "目标",
                    "values": [d["目标"] for d in self.product_data.values()],
                    "color": "primary",
                    "unit": "万元"
                },
                {
                    "name": "实际",
                    "values": [d["实际"] for d in self.product_data.values()],
                    "color": "secondary",
                    "unit": "万元"
                }
            ]
        }

        return {
            "summary": summary,
            "insights": insights,
            "chartData": chart_data,
        }

    def _query_city_data(self, message: str) -> dict:
        """查询城市数据"""
        summary = "各城市销售数据如下：\n\n"
        for name, data in self.city_data.items():
            summary += f"- {name}：销售额 {data['销售额']}万，订单量 {data['订单量']}，客单价 {data['客单价']}元\n"

        insights = [
            "上海销售额最高，达150万",
            "深圳客单价领先，为355元",
            "广州订单量相对较少，可加大推广力度"
        ]

        chart_data = {
            "title": "各城市销售额对比（万元）",
            "categories": list(self.city_data.keys()),
            "series": [
                {
                    "name": "销售额",
                    "values": [d["销售额"] for d in self.city_data.values()],
                    "color": "primary",
                    "unit": "万元"
                }
            ]
        }

        return {
            "summary": summary,
            "insights": insights,
            "chartData": chart_data,
        }

    def _query_department_data(self, message: str) -> dict:
        """查询部门数据"""
        summary = "各部门业绩情况如下：\n\n"
        for name, data in self.department_data.items():
            summary += f"- {name}：员工 {data['员工数']}人，业绩 {data['业绩']}万，人均产出 {data['人均产出']}万\n"

        insights = [
            "销售部业绩最高，达470万",
            "各部门人均产出差距不大，在17.5-18.8万之间",
            "销售部人均产出略高于其他部门"
        ]

        chart_data = {
            "title": "各部门业绩对比（万元）",
            "categories": list(self.department_data.keys()),
            "series": [
                {
                    "name": "业绩",
                    "values": [d["业绩"] for d in self.department_data.values()],
                    "color": "primary",
                    "unit": "万元"
                }
            ]
        }

        return {
            "summary": summary,
            "insights": insights,
            "chartData": chart_data,
        }

    def _default_response(self, message: str) -> dict:
        """默认回复"""
        return {
            "summary": "抱歉，我暂时无法理解您的问题。您可以尝试询问关于产品线、城市销售或部门业绩的数据。",
            "insights": [],
            "chartData": None,
        }
